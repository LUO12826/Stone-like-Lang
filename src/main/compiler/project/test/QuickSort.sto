var a = [3,4,1,2,6,7,9,5,8,10,11,16,17,19,22,23,45,18]

quickSort(a, 18)

print(a)

func partition(a, s, t) {
        var i = s, j = t, temp = a[i]
        while i < j {
            while j > i && a[j] >= temp
            {
				j=j-1
			}
            a[i] = a[j]

            while i < j && a[i] <= temp
            {
			   i=i+1
			}

            a[j] = a[i]
        }
        a[i] = temp
        return i
}

func sort(a, low, high) {
        if low < high {
            var i = partition(a, low, high)
            sort(a, 0, i - 1)
            sort(a, i + 1, high)
        }
}

func quickSort(a, length) {
        sort(a, 0, length - 1)
}
