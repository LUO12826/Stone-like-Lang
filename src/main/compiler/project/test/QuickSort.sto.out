###DATA###
NUM 1.0
NUM 1.0
NUM 1.0
NUM 0.0
NUM 1.0
NUM 1.0
NUM 0.0
ARR 18
NUM 3.0
NUM 4.0
NUM 1.0
NUM 2.0
NUM 6.0
NUM 7.0
NUM 9.0
NUM 5.0
NUM 8.0
NUM 10.0
NUM 11.0
NUM 16.0
NUM 17.0
NUM 19.0
NUM 22.0
NUM 23.0
NUM 45.0
NUM 18.0
NUM 18.0

###CODE###
j     	100	-1	LOCAL
pop   	3	-1	LOCAL
push  	3	-1	LOCAL
print 	0	-1	LOCAL
ret   	0	-1	LOCAL
pop   	3	-1	LOCAL
pop   	4	-1	LOCAL
pop   	5	-1	LOCAL
push  	4	-1	LOCAL
pop   	6	-1	LOCAL
push  	5	-1	LOCAL
pop   	7	-1	LOCAL
push  	6	-1	LOCAL
push  	3	-1	LOCAL_HEAP
pop   	8	-1	LOCAL
push  	6	-1	LOCAL
push  	7	-1	LOCAL
lt    	0	-1	LOCAL
jne   	56	-1	LOCAL
push  	7	-1	LOCAL
push  	6	-1	LOCAL
gt    	0	-1	LOCAL
push  	7	-1	LOCAL
push  	3	-1	LOCAL_HEAP
push  	8	-1	LOCAL
ge    	0	-1	LOCAL
and   	0	-1	LOCAL
jne   	33	-1	LOCAL
push  	7	-1	LOCAL
push  	0	-1	DATA
sub   	0	-1	LOCAL
pop   	7	-1	LOCAL
j     	19	-1	LOCAL
push  	7	-1	LOCAL
push  	3	-1	LOCAL_HEAP
push  	6	-1	LOCAL
pop   	3	-1	LOCAL_HEAP
push  	6	-1	LOCAL
push  	7	-1	LOCAL
lt    	0	-1	LOCAL
push  	6	-1	LOCAL
push  	3	-1	LOCAL_HEAP
push  	8	-1	LOCAL
le    	0	-1	LOCAL
and   	0	-1	LOCAL
jne   	51	-1	LOCAL
push  	6	-1	LOCAL
push  	1	-1	DATA
add   	0	-1	LOCAL
pop   	6	-1	LOCAL
j     	37	-1	LOCAL
push  	6	-1	LOCAL
push  	3	-1	LOCAL_HEAP
push  	7	-1	LOCAL
pop   	3	-1	LOCAL_HEAP
j     	15	-1	LOCAL
push  	8	-1	LOCAL
push  	6	-1	LOCAL
pop   	3	-1	LOCAL_HEAP
push  	6	-1	LOCAL
ret   	0	-1	LOCAL
pop   	3	-1	LOCAL
pop   	4	-1	LOCAL
pop   	5	-1	LOCAL
push  	4	-1	LOCAL
push  	5	-1	LOCAL
lt    	0	-1	LOCAL
jne   	89	-1	LOCAL
push  	5	-1	LOCAL
push  	4	-1	LOCAL
push  	3	-1	LOCAL
push  	3	-1	CONSTANT
call  	5	6	
pop   	6	-1	LOCAL
push  	6	-1	LOCAL
push  	2	-1	DATA
sub   	0	-1	LOCAL
push  	3	-1	DATA
push  	3	-1	LOCAL
push  	3	-1	CONSTANT
call  	61	7	
push  	5	-1	LOCAL
push  	6	-1	LOCAL
push  	4	-1	DATA
add   	0	-1	LOCAL
push  	3	-1	LOCAL
push  	3	-1	CONSTANT
call  	61	7	
j     	89	-1	LOCAL
ret   	0	-1	LOCAL
pop   	3	-1	LOCAL
pop   	4	-1	LOCAL
push  	4	-1	LOCAL
push  	5	-1	DATA
sub   	0	-1	LOCAL
push  	6	-1	DATA
push  	3	-1	LOCAL
push  	3	-1	CONSTANT
call  	61	5	
ret   	0	-1	LOCAL
push  	7	-1	DATA
pop   	0	-1	TEMP
push  	8	-1	DATA
push  	9	-1	DATA
push  	10	-1	DATA
push  	11	-1	DATA
push  	12	-1	DATA
push  	13	-1	DATA
push  	14	-1	DATA
push  	15	-1	DATA
push  	16	-1	DATA
push  	17	-1	DATA
push  	18	-1	DATA
push  	19	-1	DATA
push  	20	-1	DATA
push  	21	-1	DATA
push  	22	-1	DATA
push  	23	-1	DATA
push  	24	-1	DATA
push  	25	-1	DATA
pop   	0	17	TEMP
pop   	0	16	TEMP
pop   	0	15	TEMP
pop   	0	14	TEMP
pop   	0	13	TEMP
pop   	0	12	TEMP
pop   	0	11	TEMP
pop   	0	10	TEMP
pop   	0	9	TEMP
pop   	0	8	TEMP
pop   	0	7	TEMP
pop   	0	6	TEMP
pop   	0	5	TEMP
pop   	0	4	TEMP
pop   	0	3	TEMP
pop   	0	2	TEMP
pop   	0	1	TEMP
pop   	0	0	TEMP
push  	0	-1	TEMP
pop   	3	-1	GLOBAL
push  	26	-1	DATA
push  	3	-1	GLOBAL
push  	2	-1	CONSTANT
call  	90	4	
push  	3	-1	GLOBAL
push  	1	-1	CONSTANT
call  	1	4	
halt  	0	-1	LOCAL
